src/core/Plasma.cu

```diff
@@ -0,0 +1,1161 @@
+
+    memory_monitor("before275", nt);
+
+    checkControlPoint(275, nt);
+    SetPeriodicCurrentComponent(pd->d_CellArray, pd->d_Jx, 0, Nx, Ny, Nz);
+    SetPeriodicCurrentComponent(pd->d_CellArray, pd->d_Jy, 1, Nx, Ny, Nz);
+    SetPeriodicCurrentComponent(pd->d_CellArray, pd->d_Jz, 2, Nx, Ny, Nz);
+
+    checkControlPoint(400, nt);
+}
+
+void Plasma::AssignCellsToArraysGPU() {
+    int Nx = pd->nx, Ny = pd->ny, Nz = pd->nz;
+    int err;
+    size_t sz;
+    dim3 dimGrid(Nx, Ny, Nz), dimBlockExt(CellExtent, CellExtent, CellExtent);
+
+    err = cudaDeviceGetLimit(&sz, cudaLimitStackSize);
+    CHECK_ERROR("DEVICE LIMIT", err);
+    printf("%s:%d - stack limit %d\n", __FILE__, __LINE__, ((int) sz));
+
+    err = cudaDeviceSetLimit(cudaLimitStackSize, 64 * 1024);
+    CHECK_ERROR("DEVICE LIMIT", err);
+    printf("%s:%d - set stack limit \n", __FILE__, __LINE__);
+
+    err = cudaDeviceGetLimit(&sz, cudaLimitStackSize);
+    CHECK_ERROR("DEVICE LIMIT", err);
+    printf("%s:%d - stack limit %d \n", __FILE__, __LINE__, ((int) sz));
+
+    void *args[] = {(void *) &pd->d_CellArray, &pd->d_Ex, &pd->d_Ey, &pd->d_Ez, &pd->d_Hx, &pd->d_Hy, &pd->d_Hz, 0};
+    cudaError_t cudaStatus = cudaLaunchKernel(
+            (const void *) GPU_SetFieldsToCells, // pointer to kernel func.
+            dimGrid,                             // grid
+            dimBlockExt,                         // block
+            args,                                // arguments
+            16000,
+            0
+    );
+
+    cudaDeviceSynchronize();
+
+}
+
+void Plasma::readControlPoint(const char * fn, int num, int nt, int part_read, int field_assign,
+                              double *hx, double *hy, double *hz,
+                              double *ex, double *ey, double *ez,
+                              double *jx, double *jy, double *jz,
+                              double *qx, double *qy, double *qz
+) {
+    std::cout << fn << std::endl;
+
+    readVar(fn, "Ex", (void *) ex);
+    readVar(fn, "Ey", (void *) ey);
+    readVar(fn, "Ez", (void *) ez);
+
+    readVar(fn, "Mx", (void *) hx);
+    readVar(fn, "My", (void *) hy);
+    readVar(fn, "Mz", (void *) hz);
+
+    readVar(fn, "Jx", (void *) jx);
+    readVar(fn, "Jy", (void *) jy);
+    readVar(fn, "Jz", (void *) jz);
+
+    readVar(fn, "Qx", (void *) qx);
+    readVar(fn, "Qy", (void *) qy);
+    readVar(fn, "Qz", (void *) qz);
+
+    if (field_assign == 1) pi->AssignArraysToCells();
+}
+
+void Plasma::checkControlPoint(int num, int nt) {
+    double t_ex, t_ey, t_ez, t_hx, t_hy, t_hz, t_jx, t_jy, t_jz;
+    double t_qx, t_qy, t_qz;
+    int Nx = pd->nx, Ny = pd->ny, Nz = pd->nz;
+
+    if ((nt != TOTAL_STEPS) || (num != 600)) {
+#ifndef CONTROL_POINT_CHECK
+        return;
+#endif
+    }
+
+    memory_monitor("checkControlPoint1", nt);
+
+    if (nt % FORTRAN_NUMBER_OF_SMALL_STEPS != 0) return;
+
+    memory_monitor("checkControlPoint2", nt);
+
+    readControlPoint(pd->checkFile, num, nt, 1, 0,
```

Comment:
pd->dbgXXX is only available if macros DEBUG_PLASMA is defined. And it only makes sense if pd->checkFile is set.
I suggest not to define this macros but use pd->checkFile instead.